src/core/providers/qgsdataprovider.h

```diff
@@ -141,6 +141,11 @@ class CORE_EXPORT QgsDataProvider : public QObject
                      const QgsDataProvider::ProviderOptions &providerOptions = QgsDataProvider::ProviderOptions(),
                      QgsDataProvider::ReadFlags flags = QgsDataProvider::ReadFlags() );
 
+    /**
+     * Returns the permanent storage type for this layer as a friendly name.
```

Comment:
As far as this particular change is concerned, I'd suggest just using the label "From/to datasource database" instead of "From/to database (<storageType>)" in the load/save style dialogs, and hence removing the requirement of having `storageType` in the base QgsDataProvider. Further changes I'd postpone to a separate PR.